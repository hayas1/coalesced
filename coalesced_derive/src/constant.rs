use std::cell::LazyCell;

use syn::{parse_quote, Ident, Path};

pub const DERIVE_CONSTRUCTION: &str = "Construction";

thread_local! {
    pub static PATH_SEMIGROUP: LazyCell<Path> = LazyCell::new(|| parse_quote!(::coalesced::Semigroup));
    pub static IDENT_SEMIGROUP_OP: LazyCell<Ident> = LazyCell::new(|| parse_quote!(semigroup_op));
    pub static PATH_ANNOTATED: LazyCell<Path> = LazyCell::new(|| parse_quote!(::coalesced::Annotated));
    pub static PATH_REVERSED: LazyCell<Path> =
        LazyCell::new(|| parse_quote!(::coalesced::op::reverse::Reversed));
}
