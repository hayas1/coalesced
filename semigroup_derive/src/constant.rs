use syn::{parse_quote, Attribute, Path, TypeParam};

pub const DERIVE_CONSTRUCTION: &str = "Construction";
pub const DERIVE_SEMIGROUP: &str = "Semigroup";

#[derive(Debug, Clone)]
pub struct Constant {
    pub path_semigroup: Path,
    pub path_annotated_semigroup: Path,
    pub path_annotated: Path,
    pub path_annotate: Path,
    pub path_monoid: Path,
    pub path_commutative: Path,
    pub path_construction_trait: Path,
    pub path_construction_annotated: Path,
    pub path_construction_monoid: Path,
    pub default_type_param: TypeParam,
    pub attr_feature_monoid: Option<Attribute>,
    pub attr_feature_async: Option<Attribute>,
}
pub trait ConstantExt {
    fn constant() -> Constant;
}
pub enum External {}
impl ConstantExt for External {
    fn constant() -> Constant {
        Constant {
            path_semigroup: parse_quote! {::semigroup::Semigroup},
            path_annotated_semigroup: parse_quote! {::semigroup::AnnotatedSemigroup},
            path_annotated: parse_quote! {::semigroup::Annotated},
            path_annotate: parse_quote! {::semigroup::Annotate},
            path_monoid: parse_quote! {::semigroup::Monoid},
            path_commutative: parse_quote! {::semigroup::Commutative},
            path_construction_trait: parse_quote! {::semigroup::Construction},
            path_construction_annotated: parse_quote! {::semigroup::ConstructionAnnotated},
            path_construction_monoid: parse_quote! {::semigroup::ConstructionMonoid},
            default_type_param: parse_quote! { A },
            attr_feature_monoid: None,
            attr_feature_async: None,
        }
    }
}
pub enum Internal {}
impl ConstantExt for Internal {
    fn constant() -> Constant {
        Constant {
            path_semigroup: parse_quote! {crate::Semigroup},
            path_annotated_semigroup: parse_quote! {crate::AnnotatedSemigroup},
            path_annotated: parse_quote! {crate::Annotated},
            path_annotate: parse_quote! {crate::Annotate},
            path_monoid: parse_quote! {crate::Monoid},
            path_commutative: parse_quote! {crate::Commutative},
            path_construction_trait: parse_quote! {crate::Construction},
            path_construction_annotated: parse_quote! {crate::ConstructionAnnotated},
            path_construction_monoid: parse_quote! {crate::ConstructionMonoid},
            default_type_param: parse_quote! { A },
            attr_feature_monoid: Some(parse_quote! {#[cfg(feature = "monoid")]}),
            attr_feature_async: Some(parse_quote! {#[cfg(feature = "async")]}),
        }
    }
}
