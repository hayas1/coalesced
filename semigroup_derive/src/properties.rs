use proc_macro2::TokenStream;
use quote::ToTokens;
use syn::ItemStruct;

use crate::{
    constant::ConstantExt,
    properties::{attr::ContainerAttr, documented::Documented},
};

mod attr;
mod documented;

pub fn impl_properties<C: ConstantExt>(
    attr: &ContainerAttr,
    item: &ItemStruct,
) -> syn::Result<TokenStream> {
    let constant = C::constant();
    let documented = Documented::new(&constant, attr, item)?;
    Ok(documented.to_token_stream())
}

#[cfg(test)]
mod tests {
    use rstest::rstest;

    use crate::constant::{External, Internal};

    use super::*;

    #[rstest]
    #[case::properties_annotated(
        "properties_annotated",
        impl_properties::<External>,
        (
            syn::parse_quote! {
                annotated, monoid
            },
            syn::parse_quote! {
                /// A semigroup construction that returns the first non-`None` value.
                /// # Properties
                /// <!-- properties -->
                #[derive(Construction)]
                #[construction(annotated, monoid)]
                pub struct Coalesce<T>(pub Option<T>);
            },
        ),
    )]
    #[case::properties_not_annotated(
        "properties_not_annotated",
        impl_properties::<Internal>,
        (
            syn::parse_quote! {

            },
            syn::parse_quote! {
                /// A semigroup struct that returns the sum and overwrite
                /// # Properties
                /// <!-- properties -->
                #[derive(SemigroupInternal)]
                #[semigroup(with = "semigroup::op::Overwrite")]
                pub struct UnnamedStruct<T: std::ops::Add> (
                    #[semigroup(with = "semigroup::op::Added")]
                    T,
                    u64
                );
            },
        )
    )]
    fn test_derive_properties_snapshot(
        #[case] case: &str,
        #[case] f: impl Fn(&ContainerAttr, &ItemStruct) -> syn::Result<TokenStream>,
        #[case] (attr, input): (ContainerAttr, ItemStruct),
    ) {
        let generated = f(&attr, &input).unwrap();
        let formatted = prettyplease::unparse(&syn::parse2(generated).unwrap());
        insta::with_settings!({ snapshot_path => "../tests/snapshots", prepend_module_to_snapshot => false }, {
            insta::assert_snapshot!(case, formatted);
        });
    }
}
