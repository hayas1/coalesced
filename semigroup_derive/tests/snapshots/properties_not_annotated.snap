---
source: semigroup_derive/src/properties.rs
assertion_line: 77
expression: formatted
---
/// A semigroup struct that returns the sum and overwrite
/// # Properties
/// <!-- properties start -->
///| annotated | monoid | commutative |
///|-----------|--------|-------------|
///| ❌        | ❌     | ❌          |
///<!-- properties end -->
pub struct UnnamedStruct<T: std::ops::Add>(
    #[semigroup(with = "semigroup::op::semigroup::add::Added")]
    T,
    u64,
);
